sway-core/src/type_engine/unresolved_type_check.rs

```diff
@@ -0,0 +1,248 @@
+                resolved_type_of_parent,
+                ..
+            } => prefix
+                .check_for_unresolved_types()
+                .into_iter()
+                .chain(
+                    resolved_type_of_parent
+                        .check_for_unresolved_types()
+                        .into_iter(),
+                )
+                .collect(),
+            EnumInstantiation {
+                enum_decl,
+                contents,
+                ..
+            } => {
+                let mut buf = if let Some(contents) = contents {
+                    contents.check_for_unresolved_types().into_iter().collect()
+                } else {
+                    vec![]
+                };
+                buf.append(
+                    &mut enum_decl
+                        .variants
+                        .iter()
+                        .flat_map(|x| x.r#type.check_for_unresolved_types())
+                        .collect(),
+                );
+                buf
+            }
+            SizeOfValue { expr } => expr.check_for_unresolved_types(),
+            AbiCast { address, .. } => address.check_for_unresolved_types(),
+            // storage access can never be generic
+            StorageAccess { .. } | Literal(_) | AbiName(_) | FunctionParameter => vec![],
+        }
+    }
+}
+impl UnresolvedTypeCheck for TypedAstNode {
+    fn check_for_unresolved_types(&self) -> Vec<CompileError> {
+        self.content.check_for_unresolved_types()
+    }
+}
+impl UnresolvedTypeCheck for TypedDeclaration {
+    // this is only run on entry nodes, which must have all well-formed types
+    fn check_for_unresolved_types(&self) -> Vec<CompileError> {
+        use TypedDeclaration::*;
+        match self {
+            VariableDeclaration(decl) => {
+                let mut body = decl.body.check_for_unresolved_types();
+                body.append(&mut decl.type_ascription.check_for_unresolved_types());
+                body
+            }
+            FunctionDeclaration(decl) => {
+                let mut body: Vec<CompileError> = decl
+                    .body
+                    .contents
+                    .iter()
+                    .flat_map(UnresolvedTypeCheck::check_for_unresolved_types)
+                    .collect();
+                body.append(&mut decl.return_type.check_for_unresolved_types());
+                body.append(
+                    &mut decl
+                        .type_parameters
+                        .iter()
+                        .map(|x| &x.type_id)
+                        .flat_map(UnresolvedTypeCheck::check_for_unresolved_types)
+                        .collect(),
+                );
+                body
+            }
+            ConstantDeclaration(TypedConstantDeclaration { value, .. }) => {
+                value.check_for_unresolved_types()
+            }
+            StorageReassignment(TypeCheckedStorageReassignment { fields, rhs, .. }) => fields
```

Comment:
`StorageReassignment` cannot have generics either. I suppose this doesn't hurt though.

Reply:
It was easy enough to write...maybe we will have generic storage in the future...shrug